File: assignments/assign01/assn01_ws/src/heatControl_py/heatControl_py/thermostat.py
#!/usr/bin/env python3
import rclpy   # rclpy must be declared as a dependence in package.xml - this should be added automatically when you create the package
from rclpy.node import Node

from example_interfaces.msg import Int32
from example_interfaces.msg import String

class Thermostat(Node):
    def __init__(self):
        super().__init__("thermostat")

        # create publisher
        self.thermostatPublisher_ = self.create_publisher(
            String, 
            "fanCmd", 
            10
        )

        # create subscriber
        self.thermostatSubscriber_ = self.create_subscription(
            Int32,
            "currentTemp",
            self.callback_handleTemp,
            10
        )

        self.initialize()

    def callback_handleTemp(self, msg):
        self.temperature_ = msg.data

        if(self.temperature_ < self.lowSetPoint_):
            self.lastCmd_ = "on"
        elif(self.temperature_ > self.highSetPoint_):
            self.lastCmd_ = "off"

        # publish the cmd
        cmdMsg = String()
        cmdMsg.data = self.lastCmd_
        self.thermostatPublisher_.publish(cmdMsg)
        self.get_logger().info("the cmd is " + self.lastCmd_)


    def initialize(self):
        self.lastCmd_ = ""
        self.temperature_ = 0
        self.lowSetPoint_ = 90
        self.highSetPoint_ = 110

def main(args = None):
    rclpy.init(args=args)
    node = Thermostat()
    rclpy.spin(node)
    rclpy.shutdown()        
        
if __name__ == "__main__":
    main()  

File: assignments/assign01/assn01_ws/src/heatControl_cpp/src/thermostat.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int32.hpp"
#include "example_interfaces/msg/string.hpp"

class Thermostat : public rclcpp::Node
{
public:
   Thermostat() : Node("thermostat"), temperature_(0), lowSetPoint_(90), highSetPoint_(110), lastCmd_("")
   {
        thermostatPublisher_ = this->create_publisher<example_interfaces::msg::String>("fanCmd", 10);
        thermostatSubscriber_ = this->create_subscription<example_interfaces::msg::Int32>(
            "currentTemp", 
            10, 
            std::bind(&Thermostat::callback_handleTemp, this, std::placeholders::_1)
        );
   }

   void callback_handleTemp(const example_interfaces::msg::Int32::SharedPtr msg)
   {
        temperature_ = msg->data;
        auto cmdMsg = example_interfaces::msg::String();
        if(temperature_ < this->lowSetPoint_) {
            cmdMsg.data = "on";
        } else if(temperature_ > this->highSetPoint_) {
            cmdMsg.data = "off";
        }
        thermostatPublisher_->publish(cmdMsg);
   }

private:
   int temperature_;
   int lowSetPoint_;
   int highSetPoint_;
   std::string lastCmd_;
   rclcpp::Publisher<example_interfaces::msg::String>::SharedPtr thermostatPublisher_;
   rclcpp::Subscription<example_interfaces::msg::Int32>::SharedPtr thermostatSubscriber_;
};


int main(int argc, char **argv)
{
   rclcpp::init(argc, argv);
   auto node = std::make_shared<Thermostat>();
   rclcpp::spin(node);
   rclcpp::shutdown();
   return 0;
}